applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp

```diff
@@ -0,0 +1,655 @@
+        
+        rElement.EquationIdVector(rEquationId, rCurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function is designed to calculate just the RHS contribution
+     * @param rElement The element to compute
+     * @param RHS_Contribution The RHS vector contribution
+     * @param rEquationIdVector The ID's of the element degrees of freedom
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    void CalculateRHSContribution(
+        Element& rElement,
+        LocalSystemVectorType& rRHS_Contribution,
+        Element::EquationIdVectorType& EquationId,
+        const ProcessInfo& rCurrentProcessInfo
+        ) override
+    {
+        KRATOS_TRY
+
+        // Derivative of basis_i wrt basis_j
+        const auto basis_i = rCurrentProcessInfo[BASIS_I];
+        const auto basis_j = rCurrentProcessInfo[BASIS_J];
+
+        // Get element DOF list
+        TElementDofsVectorType r_element_dof_list;
+        rElement.GetDofList(r_element_dof_list, rCurrentProcessInfo);
+        const auto element_dofs_size = r_element_dof_list.size();
+
+        // Get PhiElemental
+        LocalSystemVectorType phi_elemental;
+        this->GetPhiElemental(phi_elemental, basis_i, rElement, r_element_dof_list, rCurrentProcessInfo);
+
+        // Compute element LHS derivative
+        Matrix element_matrix_derivative;
+        element_matrix_derivative.resize(element_dofs_size,element_dofs_size,false);       
+        if (mFiniteDifferenceTypeFlag) // Central Difference
+            this->CentralDifferencingWithBasis(element_matrix_derivative, basis_j, rElement, rCurrentProcessInfo);
+        else // Forward difference
+            this->ForwardDifferencingWithBasis(element_matrix_derivative, basis_j, rElement, rCurrentProcessInfo);
+
+        // Compute RHS contribution
+        rRHS_Contribution.resize(element_dofs_size);
+        rRHS_Contribution.clear();
+        rRHS_Contribution -= prod(element_matrix_derivative, phi_elemental);
+
+        rElement.EquationIdVector(EquationId,rCurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+    ///@}
+
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    /**
+     * @brief This function locks element nodes for finite differencing in parallel.
+     * @details 
+     * @param rElement The element to lock nodes
+     */
+    void LockElementNodes(Element& rElement)
+    {
+        KRATOS_TRY
+
+        auto& element_nodes = rElement.GetGeometry();
+        TNodePointerVectorType element_locked_nodes;
+        element_locked_nodes.reserve(element_nodes.size());
```

Comment:
I think this can be done much simpler: Simple save an index to the last node you were able to lock. In case locking all of them doesn't work you can simply iterate until this index